### PR TITLE
Add periodic job for testing certificate rotation after 1 year (bumping date in libvirt nested cluster)

### PR DESCRIPTION
**Pull request by sallyom**
This periodic test starts a nested libvirt (3 control plane nodes, 2 compute nodes) cluster in a gcp instance and:
* stop / start crio & kubelet in each node
* sets date of each node ahead
*  set date of gcp host ahead
* approves all certificate-signing-requests
* waits for:
    * all nodes Ready
    * no pending pods
    * all ClusterOperators are Available=True, Pending=False, Degraded=False

should fail until: https://github.com/openshift/cluster-kube-scheduler-operator/pull/288

### CODE REVIEW

**Comment by sallyom** [comment]
/hold

until this merges: https://github.com/openshift/release/pull/12825

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [inline ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:273]: `oc wait nodes --all --for=condition=Ready`

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [inline ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:335]: `oc wait co --all --for='condition=Available=True' && oc wait co --all --for='condition=Progressing=False' && oc wait co --all --for='condition=Degraded=False'`

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:266]: yup updating those

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:323]: yes, here the comment is misleading, it takes a few minutes for the time to propegate and to be able to reach the server - usually around 2 min from what I've seen, but I'll loop

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:256]: because with libvirt need to allow for extra time, the `wait-for install-complete` below gives extra time and/or errors

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh]: i have to check that, it's only executing for the first ip - i'll(the 1..5 was leftover from when i forgot about --no-headers) i've updated to 0..4 with --no-headers but not sure why your suggestion isn't working, leaving that for now

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:250]: hm, how to do that? looking..

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:41]: i'll test w/ that and update, I think I tried it though b4 and landed on this bc it's the only thing that worked - you're probably right though

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:33]: nope :shrug:

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:28]: i removed the gratuitous echoes : )

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh]: updated

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:2]: i'll test w/ that in a next run, leaving it for now

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:11]: oh right we can - [ it's baked into the CI image that is used to insall all libvirt envs.](https://github.com/sallyom/openshift4-libvirt-gcp/blob/master/tools/create-cluster-mirrored-local-registry) but that image needs an update to include the SAN cert - I did not want to chance breaking everyone by updating that image though so I've embedded the script here. Once it's settled and I update the image I just have to call `create-cluster-mirrored-local-registry` from inside the gcp instance

File: ci-operator/config/openshift/release/openshift-release-master__ocp-4.7.yaml
Comment (sallyom) [reply ci-operator/config/openshift/release/openshift-release-master__ocp-4.7.yaml:112]: oh right, fixed

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:13]: updated

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:350]: yea, kind of : )

**Comment by sallyom** [comment]
The image-registry CO is Available=True,Progressing=False,Degraded=True due to 
```
    message: 'ImagePrunerDegraded: Job has reached the specified backoff limit'
    reason: ImagePrunerJobFailed
    status: "True"
    type: Degraded
```
If the job is manually deleted, the CO goes to Degraded=False. 
otherwise cluster recovers after bumping date 370 days ahead (and patching the kube-scheduler-client-cert-key in the script, see here: https://github.com/openshift/cluster-kube-scheduler-operator/pull/288)

**Comment by tnozicka** [comment]
@soltysh ^ I wonder if cronjobs break if you move time ahead

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:260]: https://github.com/openshift/cluster-kube-apiserver-operator/blob/eafc4009be6dfac5d46a5ee5b203142bfb8a73ae/pkg/operator/certrotationcontroller/certrotationcontroller.go#L120-L129

https://github.com/openshift/cluster-kube-apiserver-operator/blob/df9b944038c380159bee268dc48d410294fbe70e/pkg/operator/starter.go#L216-L219

Comment (tnozicka) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:260]: `oc create -n openshift-config configmap unsupported-cert-rotation-config --from-literal='base=1d'`

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [inline ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:292]: I meant it instead of this whole function, it checks all nodes and has a timeout so I don't see the need for retry logic (simulating the timeout)

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [inline ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:325]: why is it better to retry it then raise the timeout?

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:292]: as the nodes are rebooting, csrs come in, so have to rerun the approveCSRs fn, also

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:325]: for this checkDegraded, the timeout alone will work, I'll update that.  For the checkAvailable, checkProgressing, I've found the jumpstartNodes is required to move the recovery along.

**Comment by sallyom** [comment]
Another run, same as this https://github.com/openshift/release/pull/12878#issuecomment-715672486 but updated w/ review feedback, also fails only due to image-registry Degraded=True,Available=True,Progressing=False due to image prune job:
`image-registry                             4.7.0-0.nightly-2020-10-26-152308   True        False         True       34m`

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:292]: would it need to wait for csr requests not nodes ready though?

**Comment by soltysh** [comment]
> @soltysh ^ I wonder if cronjobs break if you move time ahead

They should not, the current controller reacts periodically so it will respond properly. The only time when it will mis-behave is when the current run time aligns with your edit. In which case first one wins.

**Comment by sallyom** [comment]
@tnozicka I've removed the kube-scheduler-client-cert-key patch, let's see that this fails now, : )

**Comment by sallyom** [comment]
> /hold
> until this merges: #12825

/hold cancel

**Comment by tnozicka** [comment]
```
could not run steps: step [release:latest] failed: unable to import latest release image: imagestream.image.openshift.io "release" not found
```
/retest

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [inline ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:327]: this may have changes since now we are deleting all pods while skewing time

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [inline ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:403]: sudo, I'd just run this whole script as root from the invocation but that may be a matter of preference

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [inline ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:409]: isn't this on for the whole script already

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [inline ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:405]: nit: move it above https://github.com/openshift/release/pull/12878/files#diff-d2dafc1415630d370225f3b8a10394b14e8d13d0f5d9f7056413251863d73eeeR403 so it is logically grouped with modifying the nodes

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [inline ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:411]: there are 3 KAS whose serving certs are recover independently. Maybe just ignore the connection errors while waiting CSRs.

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [inline ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:431]: nit: I'd maybe do it using a trap fro EXIT and avoided the if

**Comment by sallyom** [comment]
@tnozicka look! 4.6 failed while 4.7 is green - that's very promising :)

**Comment by tnozicka** [comment]
@sallyom great job! We are on the right track here. Although only 2 operators are degraded while it should be any that doesn't have a static pod. The issue in our current design is that operators are regular pods and until they are started they can't update the status. I had an idea to extend clusteroperators to do a heartbeat so we can be sure the status is up to date, but never got the time. There are 2 corner cases:
- KCM doesn't start (or has cert issues), none of the operator pods are even created but it still claims Available+!Progressing+!Degraded from the past
- KS doesn't start (or has cert issues), all the pods are stuck in pending but it still claims Available+!Progressing+!Degraded from the past

Since we are already deleting the pods, can you also delete the conditions from the clusteroperators so we are sure the operators actually started reporting and this isn't stale?

Also I think we should run conformance tests as the last step. The job already takes 2 hours though, so maybe set up this short one to be required on PRs and the variant with e2e at the end as optional (so it can be triggered manually if needed) but with a periodic to make sure we don't regress?

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [inline ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:387]:
```suggestion

# Delete all clusteroperator status to avoid stale status when the operator pod isn't started.
export bearer=$( ioc -n openshift-cluster-version serviceaccounts get-token default ) && export server=$( oc whoami --show-server ) && for co in $( ioc get co --template='{{ range .items }}{{ printf "%s\n" .metadata.name }}{{ end }}' ); do curl -X PATCH -H "Authorization: Bearer ${bearer}" -H "Accept: application/json" -H "Content-Type: application/merge-patch+json" ${server}/apis/config.openshift.io/v1/clusteroperators/${co}/status -d '{"status": null}' -k && echo; done
```

Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:387]: added

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:404]: If i recall, I had to start kubelet _after_ setting time on the host? I might be wrong, I'll give it a try

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:408]: updated

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:326]: last i checked that was blocking recovery, but right, I didn't check after deleting all pods, I'll confirm

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:407]: you recall correctly, the line above is not doing that though. it sets the time on the bastion or wherever your run the script from, not on the master with the kubelet.

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:387]: 4.7 run passed, 4.6 run failed, with the above addition so that's good

**Comment by sallyom** [comment]
@tnozicka rebased/cleared conflicts - next week let's discuss & merge if possible :)

File: ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh
Comment (tnozicka) [inline ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh:289]: you should probably read the number on start, looks like there is just one worker now https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/12878/rehearse-12878-periodic-ci-openshift-release-master-ocp-4.7-e2e-gcp-libvirt-cert-rotation/1352747737774297088#1:build-log.txt%3A68

Comment (sallyom) [reply ci-operator/step-registry/openshift/e2e/gcp/libvirt/cert-rotation/conf/openshift-e2e-gcp-libvirt-cert-rotation-conf-commands.sh]: that was a flake, for sure, I haven't seen another run only get 1 worker.. i'll keep an eye out for it, once this merges, I'll take ownership to watch flakes

**Comment by tnozicka** [comment]
/approve
This is ready and the nits can be addressed as followups. Good job @sallyom!
Expectation is to see:
- 4.7 passing

**Comment by sallyom** [comment]
yay, rehearsals are green again. will retest to get a few good runs

/retest

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
